### PR TITLE
HCAL TP for Plan 1

### PR DESCRIPTION
**Pull request by matz-e**
Contains several previous PR, fixed:

* #17523, this PR supercedes that one.
* #17469 with enum `TriggerMode` meaning of the enumeration values unchanged.
* #17260, which got reverted with one of the L1T PR that was reverted.

@davidlange6 I fixed the RelVal 10826.0 by keeping the meaning of the trigger mode consistent, I had added the new 2016 and Plan 1 definition and changed the underlying values, which had already been saved in the geometry DB.

### CODE REVIEW

**Comment by civanch** [comment]
please test

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc [hunk 1]
```diff
@@ -204,12 +204,29 @@ CaloTPGTranscoderULUT::getOutputLUTSize(const HcalTrigTowerDetId& id) const
       case HcalTopologyMode::TriggerMode_2016:
          return QIE8_OUTPUT_LUT_SIZE;
       case HcalTopologyMode::TriggerMode_2017:
+         if (id.ietaAbs() <= theTopology->lastHERing())
+            return QIE8_OUTPUT_LUT_SIZE;
+         else
+            return QIE10_OUTPUT_LUT_SIZE;
+      case HcalTopologyMode::TriggerMode_2017plan1:
+         if (theTopology->dddConstants()->isPlan1(id))
+            return QIE11_OUTPUT_LUT_SIZE;
```
Comment (franzoni) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: hello @matz-e 
detIdSp_ https://github.com/matz-e/cmssw/blob/9c7c69d4203cf4d932921a4d4103956b8dc7cfbd/Geometry/HcalCommonData/interface/HcalDDDRecConstants.h#L136   is empty by construction, for now
which in turns means isPlan1(...) always returns false.

Comment (matz-e) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: Doesn't it get filled [here](https://github.com/cms-sw/cmssw/blob/a85956d574ec10a14b2dba8a1d503d4ddd0bb208/Geometry/HcalCommonData/src/HcalDDDRecConstants.cc#L896)?

**Comment by matz-e** [comment]
I've noticed that this breaks various era-conditionals and will update with a fix later today.

**Comment by kpedro88** [comment]
@matz-e ready to test?

**Comment by matz-e** [comment]
Should be ready. I should note that Salavat's simple pgun config still fails for me. At least there should be no regression.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by franzoni** [comment]
+1

File: CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc [hunk 2]
```diff
@@ -238,126 +238,123 @@ void HcaluLUTTPGCoder::update(const HcalDbService& conditions) {
...
-                   upgradeQIE10LUT_[lutId][adc] = std::min(std::max(0,int((adc2fC - ped) * gain * rcalib / lsb_ / cosh_ieta[abs(ieta)] )), QIE10_LUT_BITMASK);
-             }
-	  } // endif HF
-
-	} // for depth
-      } // for iphi
-    } // for iphi
-  }// for subdet
+  for (const auto& id: metadata->getAllChannels()) {
+     if (not (id.det() == DetId::Hcal and topo_->valid(id)))
```
Comment (davidlange6) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: is this if statement ever false in normal conditions? Eg, if its true is that a sign of corrupt data and/or incomplete conditions?

Comment (matz-e) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: When I implemented this, there were IDs that were either invalid HCAL IDs or not allowed by the topology… I'm not sure anymore which IDs exactly are skipped by this.

Comment (ianna) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: indeed there are such IDs. Verification of validity by topology supposed to be a temporary work-around, I think.

Comment (davidlange6) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: ok, thanks - is a dangerous condition to leave around, so lets try to clean up after than plan1 issues are behind us

**Comment by davidlange6** [comment]
+1

File: CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc (hunk 2 above)
Comment (rekovic) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: @matz-e Was this not the fix when simHCALTriggerPrimitives in sequence for Phase2?

Comment (matz-e) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: Yes, that was included in there, too. It somehow got reverted with the reversal of one of the L1T PRs, I think.

**Pull request merged**